Support a max point count in serializing series

djmitche opened the pull request:
### What does this PR do?

This adds a new configuration value, "serializer_max_series_points_per_payload", defining the maximum number of series points to put in a single payload.  It then finishes a payload if the next series would put it over the limit.

### Motivation

Request from intake team

### Possible Drawbacks / Trade-offs

This was a minimal change to accomplish the goal without adding unnecessary risk.

Series serialization now has three limits for a payload: compressed size, uncompressed size, and number of points.  The first two are handled by the compressor, and in this patch the third is handled by the series marshalSplitCompress function.  Note that the compressor is generic to multiple metric types, not all of which have points.

A more invasive approach would involve refactoring all of this to abstract the serialization loop currently duplicated between the different metric types, and adding support for multiple per-payload maxima.  That would constitute a change to multiple metric types, though, and has a high likelihood of introducing bugs.

### Describe how to test/QA your changes

With the previous agent version, observe that the rate at which 
```
curl http://localhost:5000/debug/vars | jq .sketch_series.PayloadFull
```
is increasing.  Alternately, look at datadog.agent.sketch_series.payload_full for the host.

Upgrade the agent, and confirm that this rate is the same.

Configure "serializer_max_series_points_per_payload" to a small value in a situation where lots of series metrics are transmitted (such as the benchmark environment).  Confirm that the rate increases (more full payloads).

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

djmitche committed:
Support a max point count in serializing series

remicalixte commented:
Is it possible to set the limit per endpoint? We'll need this new limit for the v2 api but it would be unnecessary to affect the v1 api too. Or maybe it is not a problem?

djmitche commented:
This is only for v2.  My hope is to drop v1 support once you're happy with v2.

ogaca-dd commented on pkg/metrics/series.go:
```diff
@@ -388,8 +394,18 @@ func marshalSplitCompress(iterator serieIterator, bufferContext *marshaler.Buffe
 
-		// Compress the protobuf metadata and the marshaled series
-		err = compressor.AddItem(buf.Bytes())
+		if pointsThisPayload > maxPointsPerPayload {
```
Consider the following code 

```go
func TestMarshalSplitCompressPointsLimitOne(t *testing.T) {
	mockConfig := config.Mock()
	oldMax := mockConfig.GetInt("serializer_max_series_points_per_payload")
	defer mockConfig.Set("serializer_max_series_points_per_payload", oldMax)
	mockConfig.Set("serializer_max_series_points_per_payload", 1)

	series := makeSeries(1, 2)
	payloads, err := series.MarshalSplitCompress(marshaler.DefaultBufferContext())
	require.NoError(t, err)
	require.Len(t, payloads, 0)
}
```
A serie with 2 points is created. As `serializer_max_series_points_per_payload` is `1`, the serie should be dropped (`stream.ErrItemTooBig`) whereas `MarshalSplitCompress` returns 2 payloads. Is it expected?

djmitche replied:
Good point, and thanks for the test case!  I've copied the test case and fixed the underlying issue (and two more).

djmitche committed:
Fix handling of series with too many points.

This adds a new test case provided by @ogaca-dd, and fixes the bug
causing it to fail.  The bug was using `pointsThisPayload` where I should
have used `len(serie.Points)`.

Having fixed that, the code was still flushing an empty payload when it
stopped, so the seriesThisPayload variable tracks that and skips
flushing when it is zero.

In the process of making that change, a third bug became apparent: when
the Compressor refuses a series, the existing code was still "counting"
the points in `pointsThisPayload`.  This would have had little effect,
but is remedied by only incrementing `pointsThisPayload` and
`seriesThisPayload` when the compressor accepts a series.

ogaca-dd commented on pkg/metrics/series.go:
```diff
@@ -433,9 +461,12 @@ func marshalSplitCompress(iterator serieIterator, bufferContext *marshaler.Buffe
-	if err != nil {
-		return nil, err
+	// if the last payload has any data, flush it
+	if seriesThisPayload > 0 {
```
What do you think about moving the condition `if seriesThisPayload > 0` into `finishPayload`, so that `finishPayload` creates a payload only if it is not empty? It will not change the behavior but I think it is more natural that `finishPayload` handles this check.

djmitche replied:
Sounds good.

djmitche committed:
only check seriesThisPayload in finishPayload

remicalixte commented:
Is it possible to set a lower default value for the limit? `100000` is very high. We are aiming more at something like `10000`.

djmitche commented:
From slack:
> For the limit I'm thinking of 10000 points max maybe. I'll need to run some experiments

so it looks like I added a zero -- sorry! :)
